Week1 [step 3] malrang

malrang-malrang opened the pull request:
@Tak95  어찌어찌 완성했습니다..!

스텝3를 진행하면서 가볍게 풀수있는부분 들 마저도 너무 어렵게 생각하고 삽질을 많이 했던것같습니다..
잘하고자 하는 욕심에 함수에 정보를 입력하면 정보도 출력하고 정보의 요청대로 빼빼로도 
만들어주는 함수를 만들고 싶었습니다!
중간중간 모르는것이있으면 찾아보고 공부하며 완성했으나 욕심이 너무큰탓인지 가독성이 떨어지는 코드 인것같습니다

요구사항대로 함수마다 매개변수를 받아 입력된값을토대로 기능하는 함수들로 만들수도있었으나
이러한 방식으로 구현되게끔 만들고싶었습니다..ㅠ
앞으로는 요구사항대로 먼저 만들어본뒤에 개선하거나 추가 기능을 구현하는 방법으로 스텝을 진행해야할것같습니다!

# 궁금했으나 해결된점!
``` 
For 문 을 이용해 몸통과 토핑을 몸통의 길이만큼 출력하는 코드를 작성할때
For _ in 1…peperoBodyLength {
    If peperoBody == “***”{
         If _  in % 2 == 1{
              If{
                   Print(“\(Body), \(Topping)”)
}
              Else if{
                    print(“\(Topping),\(Body)”)
}
}
}
}
```
홀수,짝수를 가정하여 작성한 코드인데 저렇게 작성하게되면 홀수에 해당하는 부분만 
반복해서 몸통의 길이만큼 출력해주어
한참을 고민하고 해맸습니다 ㅠ

그후 for 다음 언더바가 아니라 count	로 대체하고나니 그 순서를 count로 저장해 
if 에서  count가 홀수인진 짝수인지 묻게끔 만들었더니 해결되었습니다!

Commit:
첫 커밋

Commit:
문제1답변

Commit:
문제1,2,답변

Commit:
1번문제 답

Commit:
1,2번 문제답

Commit:
1,2,번문제답

Commit:
step1수정본

Commit:
step2project,순서도

Commit:
step2수정본

Commit:
for->while,세부사항수정.

Commit:
규칙을 가지 고있는 함수를 반복하는 함수로 지정,주석제거

Commit:
규칙을 가지 고있는 함수를 반복하는 함수로 지정수정본

Commit:
<label>:<title>

<issue-number>

Commit:
Feat.step3

5nTak commented on CodeStarterCamp_Week1/1week step3.swift.swift:
```diff
@@ -0,0 +1,67 @@
+var peperoBody : String = ""
+var peperoTopping : String = ""
+
+func makePeperoInfo(BarLength : Int, BodyLength : Int, Body : String, Topping : String){
```
매개변수를 활용하신 점 너무 좋았습니다! ㅎㅎ
근데 매개변수를 UpperCamelCase로 선언하셨네요?! 특별히 의도하신게 있을까요?

malrang-malrang replied:
앗! 네이밍 가이드를 보며 이리저리 수정하다보니 마지막에 확인을 못한것같습니다 ㅠㅠ
원래는 중간에있는 단어였답니다..하하..!
수정하겠습니다!

5nTak replied:
함수의 선언부가 바뀌었으니 호출부도 바뀌어야 맞지 않을까요??
이대로면 실행이 안될 것 같은데 혹시 실행이 가능했나요?!

추가로 이런 경우를 방지하기 위해 네이밍을 바꿀 때는 바꿀 객체에 `마우스 우클릭` - `Refactor` - `Rename`을 하게되면 프로젝트 전체 내부에서 사용된 해당 객체의 이름을 한꺼번에 바꿀 수 있답니다!😱

malrang-malrang replied:
앗! 자꾸 깜빡깜박 하는군요 ㅠㅠ 꼼꼼히 돌아보는 습관이 필요한것같습니다 ㅠㅠ 
말씀해주신 Refactor 완전 꿀팁이네요..!

5nTak commented on CodeStarterCamp_Week1/1week step3.swift.swift:
```diff
@@ -0,0 +1,67 @@
+    
+    makePepero()
+}
+func makePepero(){
```
다른 부분들은 함수의 `}`괄호와 다음 함수 사이에 개행이 있는데 이 곳엔 개행이 없네요??
말랑님만의 코드스타일을 정해서 통일성을 지켜주세요!😃

malrang-malrang replied:
앗.! 네! 지금 다시 보니 다른곳에는 개행이 있었는데 한곳에만 없으니 살짝 불안한 느낌이 듭니다..! 정렬되지 않은느낌,,..! 
수정하겠습니다 ㅎㅎ

5nTak commented on CodeStarterCamp_Week1/1week step3.swift.swift:
```diff
@@ -0,0 +1,67 @@
+var peperoBar : Int = 0
+var peperoBodyLength : Int = 0
+var peperoBody : String = ""
+var peperoTopping : String = ""
```
코드를 살펴봤을 때, 이 변수들은 굳이 전역변수로 해야할 필요가 있었을까? 생각이 드는데 어떻게 생각하시나요?!

malrang-malrang replied:
처음 생각할 당시 빼빼로의 정보들을 전역변수를 선언해서 저장해두고 
그후 만들게될 함수들은 전역변수의 값을 이용해 함수를 만들어 
작동시킬 생각으로 만들게 되었으나

다만들고 나서는 굳이 전역변수를 선언하지않아도 구현이 가능했겠구나 하고 깨달았습니다..!
아무래도 잠깐이나마 공부했던 함수지향형 언어가아닌 절차지향언어인c언어 공부를 했던게 생각이 자리잡아 자연스럽게 
저렇게 생각하지 않았나 생각합니다..

이부분에대해 질문을 드리고싶었습니다! 
만드는 과정중 엘렌에게 코드리뷰를 받았던 적이 있었는데 전역변수를 두는것을 탁님 처럼 물어봤던 적이 있었습니다!
그때당시는 일단은 계획한대로 만들고난후 다시한번 돌이켜보자 라는 생각으로 완성시킨 결과물입니다! 
swift 에서는 일반적으로 전역변수를 사용하는게 드문 상황인건지 궁금했습니다!

5nTak replied:
이 부분은 말랑님이 조금 더 구글링을 통해 공부해보신 다음에 그 이후에도 의문이 풀리지 않는다면 다시 질문해보는걸로 하는게 어떨까요?! 😄 

현재는 command line tool로 코드를 작성하고 있지만 추후 iOS에 대해 본격적으로 프로젝트를 시작하시거나 이후 공부하실 class, struct, enum 등등 여러 타입에 대해서도 공부하다보면 자연스레 알게 될 것 같습니다! 😃

malrang-malrang replied:
네! 지금시점으로는 아무래도 함수지향 패러다임을 가지고있다보니 전역변수보다는 클래스,구조체,열거형 등 한곳에 묶어두어 저장해두는게 
보기에도 깔끔하게 사용될것같습니다! 아무래도 지금은 코드를 작성하는 페이지가 적다보니 괜찮지만 사용해야하는 전역변수들이 많이 선언되어있게되면 전역변수의 갯수도 너무많아지고 하나하나 어떤것들을 저장하는건지도 햇갈릴수 있을것같습니다 비슷한것들은 한곳에 묶어두어야 보기편하고 사용할때도 어떤기능인지 어떤것을 저장하는건지 이해가 쉬울것같습니다 ! 조금더 공부해보겠습니다!

5nTak commented on CodeStarterCamp_Week1/1week step3.swift.swift:
```diff
@@ -0,0 +1,67 @@
+          print("정보를 잘못 입력하였습니다 다시 입력해주세요")
+      }
+      }
+ }
```
기능구현은 정말 잘하셨으나, 함수가 아직 많이 무거워보여요.
for문 안에 if문, if문 안에 if문, 그뒤에 else if ... 이렇게 이어지다보니까 코드가 많이 난잡해보이죠?
이렇게 조건이 여러개일 경우에는 코드를 어떻게 짜야 더 효율적일까요?🤔

malrang-malrang replied:
네 ㅠㅠ 가독성도 많이 떨어지고 다른사람이보기에 한번에 이해하기 어려운 코드라고 생각됩니다 ㅠ
지금 드는 생각으로는 if문과 switch 를 같이 사용하면 조금은더 보기 편하지 않았을까 하는생각이듭니다
if 문안에 if 문을 넣게되니 보기에 많이 어려워 보입니다...

5nTak replied:
switch문을 생각해보셨군요! 그렇다면 한번 switch문을 사용해서 적용해볼까요?!😁

malrang-malrang replied:
네! switch 를 이용해 변경해보겠습니다! 이것보다도 더깔끔하게 복잡하지 않게 할수있는 방법이있을것같은데 
지금 단계에서는 다른 좋은 수단이 떠오르지 않네요 ㅠㅠ if 문안에 if 는 보기 불편하니...switch 로변경해보겠습니다!

5nTak commented on CodeStarterCamp_Week1/1week step3.swift.swift:
```diff
@@ -0,0 +1,67 @@
+      }
+      else {
+          print("정보를 잘못 입력하였습니다 다시 입력해주세요")
+      }
```
사용자로부터 입력받지 않았으면 하는 경우의 수도 생각해보셨군요!😁

malrang-malrang replied:
네..! 구매자가 정보를 입력하면 그것에맞게 정보가출력되고 정보대로 만들어진 빼빼로를 받게된다! 라고 생각하며 만들었더니
구매자가 옵션을 잘못선택한 경우를 가정하여 넣게되었습니다ㅋㅋㅋ ㅎㅎ

5nTak replied:
좋은 접근인 것 같습니다! 😀
다만 현재 body에 예외값을 넣어 에러를 출력해보면
<img width="273" src="https://user-images.githubusercontent.com/77046882/146263333-aebce584-6b52-431c-831a-a4a88f52e286.png">
조금 더 깔끔한 방법이 있을 것 같긴 한데. 이건 인지만 한채로 호기심으로 남겨두고 나중에 공부해보면 좋을 것 같습니다!😀

malrang-malrang replied:
앗..! 이부분에 대해서는 생각 하지못했습니다! 몸통의 길이만큼 출력되다보니 4번이나 출력 되게 되는군요..!! 그후 막대기는 출력이되다니..
잘못입력했을시에는 예외처리하여 문구만 한번 출력되게끔 바꿀수있을것같습니다!

Commit:
chore: 개항관련,네이밍 수정

통일성있게 함수와 함수사이의 개항을 맞춰 수정했습니다
UperCamelCase 로 선언된 매개변수 이름을 LowerCamelCase 로 변경하였습니다

Commit:
Fix : makeBodyShape 함수내부 수정

if 문속에 if 문이 작성되어 알아보기 힘든 코드를
switch 속에 if 문으로 수정하여 조금은 더 보기편한 코드로 수정하였습니다

5nTak commented on CodeStarterCamp_Week1/1week step3.swift.swift:
```diff
@@ -36,33 +36,31 @@ func makeBar(){
-          }
-      else if peperoBody == "| |"{
+          
+      case "| |" :
```
Misson Contents의 실행예시에는 몸통이 `|0|`로 출력되어 있던데, `| |`로 하신건 의도하신건가요??

malrang-malrang replied:
앗..! 네! 0 은 토핑 부분이라고 생각해서 몸통이 누드빼빼로라면  기본 몸통은 | | 으로 설정되고 누드빼빼로에도 
"***" 처럼 토핑을 선택하면 0말고 다른 토핑이 들어갈수있게끔 추후 변경도 가능하지않을까 해서 저렇게 작성했습니다!
현재 코드는 몸통을 누드 빼빼로로 정하게되면 토핑을 결정할수없게끔 되어있습니다..!

5nTak commented on CodeStarterCamp_Week1/1week step3.swift.swift:
```diff
@@ -0,0 +1,67 @@
+        }
+        else {
+            print("  | |")
+        }
```
이 코드대로라면 `body`가 `"***"`일때만 앞 공백이 하나 있고 그 외엔 전부 공백이 두 개 들어가게 되네요.
Misson Contents를 살펴보면 누드빼빼로의 좌측에는 빈칸이 하나 추가되어있다고 하는데, 그렇다면 **if문에는 body가 누드빼빼로일때의 조건을 걸고 else문에 그 외일 때의 조건**을 거는게 추후 확장성을 고려했을 때 더 좋을 것 같은데 어떻게 생각하시나요?!

malrang-malrang replied:
네 맞습니다! 추후 확장성을 고려한다고 하면 누드 빼뺴로를 예외로두어 고정시켜두는게 확장시킬때 더 편할것 같습니다! 수정 해보겠습니다! ㅎㅎ

5nTak commented on CodeStarterCamp_Week1/1week step3.swift.swift:
```diff
@@ -36,33 +36,31 @@ func makeBar(){
           print("정보를 잘못 입력하였습니다 다시 입력해주세요")
       }
       }
  }
```
들여쓰기가 조금 이상한 것 같죠?
이럴 때는 수정하고 싶은 영역을 드래그 한뒤 `control(^) + i` 를 눌러보세요!

5nTak commented on CodeStarterCamp_Week1/1week step3.swift.swift:
```diff
@@ -36,33 +36,31 @@ func makeBar(){
+                  }
+                  else{
+                      print(" \(peperoBody)\(peperoTopping)")
+                  }
```
switch문 안에도 if문이 들어갔네요..😅 현재 코드를 보면 
```
for문 {
    switch문 {
        if문 {
        } else {
            if문 {
            } else {
            }
        }
    }
}
```
이 상태인데요...🥲
if문의 중첩이 많아서 보기가 좋지 않으니 억지로 if문의 중첩만을 제거하려고 해서 생긴 문제 같아요.
결과적으로는 if문의 중첩은 사라지지 않고 switch문이 추가되어 코드가 더 복잡해져버렸죠💦
현재 `makeBodyShape()` 이라는 함수가 어떤 동작들을 하고있는지 정리하신 후에 해당 기능들을 분리할 수 있는데 까지 분리해보고 그 분리된 기능대로 함수를 다시 만들어 보는건 어떨까요?!
일종의 재설계를 한다? 라고 이해하시면 될 것 같아요!

malrang-malrang replied:
전역변수를 선언하지 않고 함수에서 값을 반환하여 다른함수에 넣어주는것을 못하고있어서 함수를 분리시키지못해 
함수하나에 원하는 기능을 다 넣으려다보니 이런일이 생기는것 같습니다!ㅠ 아무래도 기능들을 분리해서  세분화해야할것같습니다!

Commit:
Refactor: 빼빼로 만들기 구현 방법 변경

전역변수를 사용하지않게 변경
함수내부에 여러가지 기능이 들어있어 기능들을 나누어 따로 함수화하여 분리 시킴

5nTak commented on CodeStarterCamp_Week1/repeatStep3.swift:
```diff
@@ -0,0 +1,57 @@
+    func makeBody(body: String, topping: String) -> String {
+        switch body {
+        case "|0|" :
+            return " \(body)\(topping)\n, \(topping)| |\n"
```
콘솔앱을 다시 실행해보고 이상한 점을 찾아서 반환하는 문자열을 수정해볼까요?!

malrang-malrang replied:
앗! 아무래도 푸쉬가 잘못된것같습니다!
 저렇게바꿔보고 어떻게 출력이되는지 확인후 커밋 푸쉬를 했는데 과정중 add 하여 스테이징 공간에 넣는 과정을 생략하고 푸쉬해서
최종 수정본말고 다른게 푸쉬된것같습니다..!

5nTak commented on CodeStarterCamp_Week1/repeatStep3.swift:
```diff
@@ -0,0 +1,57 @@
+    
+    //빼빼로 몸통 만드는 함수
+    func makeBody(body: String, topping: String) -> String {
+        switch body {
```
switch문을 사용하셨는데 현재 코드에는 조건이 두가지 경우 밖에 없죠? `body가 "|0|"인지 아닌지 에 대한 조건`뿐인 것 같은데 이럴 경우엔 `if {} else {}` 문을 쓰는건 어땠을까 생각해봅니다 ! :)
switch문은 다중조건일 때 더 유용하게 쓸 수 있을 것 같아요 😄

malrang-malrang replied:
그렇군요! 추후 몸통 조건에 다른 경우를 넣어 빼뺴로 가짓수를 늘릴수있지 않을까..하여 switch 를 선택했습니다! 
if로 수정하겠습니다 :)

5nTak commented on CodeStarterCamp_Week1/repeatStep3.swift:
```diff
@@ -0,0 +1,57 @@
+        }
+    }
+    
+    //빼빼로 몸통 만드는 함수
```
코드로 충분한 설명이 이뤄졌기 때문에 주석은 불필요해보입니다😉

malrang-malrang replied:
네! 수정하겠습니다 :)

5nTak commented on CodeStarterCamp_Week1/repeatStep3.swift:
```diff
@@ -0,0 +1,57 @@
+        }
+    }
+    
+    
```
불필요한 공백!

malrang-malrang replied:
으아아!

5nTak commented on CodeStarterCamp_Week1/repeatStep3.swift:
```diff
@@ -0,0 +1,57 @@
+
+
+
+
```
여기두요!

malrang-malrang replied:
으아!

5nTak commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -41,7 +41,7 @@ func makeIceCreamBar(){
-//몸통부분은 "***", "| |" 만 선택가능합니다
-//누드빼빼로는 토핑선택이 불가능합니다
-makePeperoInfo(barLength: 2, bodyLength: 4, body: "***", topping: "")
+//몸통부분은 "***", "|0|" 만 선택가능합니다
```
주석에서 설명하기로 몸통부분은 `"***"` 와 `"|0|"`만 선택이 가능하다고 했는데 코드에는 그런 제한조건이 없는 것 같아요!🥲
몸통부분에 `해당 조건에 맞지 않는 body`가 들어왔을 때의 제한 조건을 로직에 추가해볼까요?!

malrang-malrang replied:
네! 추가하였습니다! if 를 사용해 추가 조건을 넣어주었는데 길이는 짧지만 추가조건을 넣음으로써 가독성이 조금 떨어지게된것같습니다 ㅠ
그렇다고 함수 구조자체를 바꾸자니 전처럼 if 안에 if 를 사용하는 형태가 될것같아 현재상태로 푸쉬하였습니다..!

5nTak commented on CodeStarterCamp_Week1/main.swift:
```diff
@@ -24,15 +24,23 @@ func drawIceCreamBar(){
+// makeIceCream()
+// makeIceCreamBar()
+
+
```
- 해당 주석이 필요할까요?

- 불필요한 공백!

Commit:
Feat: 잘못된 정보를 입력시 번환값 추가

기존 빼빼로 만들기 몸통을 만드는 함수에서 잘못된 바디를 선택하게되면 "잘못된 정보를 입력하였습니다" 를 출력 하게끔 수정됨

5nTak commented on CodeStarterCamp_Week1/repeatStep3.swift:
```diff
@@ -21,28 +21,34 @@ func makePrinrtInfo(barLength : Int = 4, bodyLength : Int = 10, body : String =
         let bodyArray: Array<String> = peperoBody.components(separatedBy: "\n")
-        for count in 1...bodyLength {
-            print(bodyArray[count % 2])
+        if makeBody(body: body, topping: topping) == "잘못된 정보" {
```
실행 상의 문제는 없지만 이 경우엔 `if makeBody() == "잘못된 정보"` 보단 `if peperoBody == "잘못된 정보"` 가 조금 더 적절하지 않을까요?! 우린 이미 `peperoBody`안에 `makeBody`함수의 반환값을 넣어줬잖아요?!

malrang-malrang replied:
그렇군요! 지금보니 함수에 반환된값이 "잘못된정보" 보다는 peperobody가 좀더 직관적일것같습니다!

5nTak commented on CodeStarterCamp_Week1/repeatStep3.swift:
```diff
@@ -21,28 +21,34 @@ func makePrinrtInfo(barLength : Int = 4, bodyLength : Int = 10, body : String =
             return " \(body)\(topping)\n\(topping)\(body)"
+        } else {
+            return "잘못된 정보"
         }
```
이렇게 조건이 3가지 이상이 됐을 경우 switch문을 잘 활용해볼 수 있지 않을까? 라는 의미였습니다! :)

malrang-malrang replied:
아! 이해했습니다! 1,2가지정도는 if 가 간단 하지만
3가지 이상 정도부터는 switch가 case 로나뉘기때문에 보기가 더 편할것같습니다!

Commit:
Feat: 주석및공백제거,함수내부 if에서 switch로 변경
